### PR TITLE
fix!: keep ProtoNode deserialised state stable until explicit mutation

### PR DESCRIPTION
**Pull request by rvagg**
When decoding a badly serialised block with Links out of order, don't sort
the list until we receive an explicit mutation operation. This ensures stable
DAG traversal ordering based on the links as they appear in the serialised
form and removes surprise-sorting when performing certain operations that
wouldn't be expected to mutate.

The pre-v0.4.0 behaviour was to always sort, but this behaviour wasn't baked
in to the dag-pb spec and wasn't built into go-codec-dagpb which now forms the
backend of ProtoNode, although remnants of sorting remain in some operations.
Almost all CAR-from-DAG creation code in Go uses go-codec-dagpb and
go-ipld-prime's traversal engine. However this can result in a different order
when encountering badly encoded blocks (unsorted Links) where certain
intermediate operations are performed on the ProtoNode prior to obtaining the
`Links()` list (`Links()` itself doesn't sort, but e.g. `RawData()` does).

The included "TestLinkSorting/decode" test is the only case that passes without
this patch.

Ref: https://github.com/ipld/ipld/pull/233
Ref: https://github.com/filecoin-project/boost/issues/673
Ref: https://github.com/filecoin-project/boost/pull/675

---

Further notes:

* Yes, in an ideal world we could rewind a couple of years and bake sort-on-decode into the dag-pb spec and go-codec-dagpb, but that's not the case and our most important stable-DAG-traversals are happening in CARs now which are almost exclusively using the new stack.
* Our JavaScript codecs don't sort-on-decode.
* All old and new Go and JavaScript codecs in the ipfs and ipld orgs properly sort on encode. However, there now exist multiple implementations of dag-pb that do not follow this part of the spec, and we're now encountering them in places where stable-DAG-traversals really matter.
* This is really a continuation of the ipld-in-ipfs work that lead to v0.4.0, we're just fine tuning a weird corner case.

### CODE REVIEW

**Review by willscott — APPROVED** [review]

File: node.go [hunk 1]
```diff
@@ -358,14 +385,19 @@ func (n *ProtoNode) Multihash() mh.Multihash {
...
 }
 
-// Links returns the node links.
+// Links returns a copy of the node's links.
 func (n *ProtoNode) Links() []*format.Link {
-	return n.links
+	links := make([]*format.Link, len(n.links))
+	copy(links, n.links)
+	return links
 }
```
Comment (aschmahmann) [inline node.go:393]: This seems more reasonable. Hopefully no one is using mutation here, did you check for usage here or mostly assuming folks haven't been doing anything nuts here?

File: node.go [hunk 2]
```diff
@@ -129,19 +138,29 @@ func (n *ProtoNode) AddNodeLink(name string, that format.Node) error {
...
 	})
-
+	sort.Stable(LinkSlice(n.links))
 	return nil
 }
 
-// RemoveNodeLink removes a link on this node by the given name.
+// RemoveNodeLink removes a link on this node by the given name. If there are
+// no links with this name, ErrLinkNotFound will be returned. If there are more
+// thank one link with this name, they will all be removed.
```
Comment (aschmahmann) [inline node.go:163]:
```suggestion
// than one link with this name, they will all be removed.
```

File: node.go [hunk 3]
```diff
@@ -129,19 +138,29 @@ func (n *ProtoNode) AddNodeLink(name string, that format.Node) error {
...
+// form as the links will have been sorted.
 func (n *ProtoNode) AddRawLink(name string, l *format.Link) error {
 	n.encoded = nil
 	n.links = append(n.links, &format.Link{
 		Name: name,
 		Size: l.Size,
 		Cid:  l.Cid,
 	})
-
+	sort.Stable(LinkSlice(n.links))
```
Comment (aschmahmann) [inline node.go:157]: Are we concerned about performance hits here during node creation since we're sorting after every link addition?

Would it make more sense to basically mark the node as dirty and then sort + clear the dirty bit the first time anyone encodes or tries to inspect the links?

**Review by aschmahmann — CHANGES_REQUESTED** [review]
I would also bubble this change up using a commit hash into kubo. There are a bunch of tests and history there such that if a test breaks it should link to what the original concern/issue was.

code archaeology tip: If you're curious why something is how it is you can use git blame and when you land on a commit for something that seems pretty old copy-paste the commit title (or hash) and search for it in kubo to find the related PR.

File: node.go (hunk 1 above)
Comment (rvagg) [reply node.go:393]: No, I haven't checked very broadly and I haven't seen anyone doing anything weird but I don't want to expose this if we're taking so much care with statefulness of sorting internally. But, I had to do this here https://github.com/filecoin-project/boost/pull/675 to deal with the inverse - that go-merkledag was messing with it.

File: node.go (hunk 3 above)
Comment (rvagg) [reply node.go]: No, I wasn't too concerned about performance because this is mostly an insertion-into-sorted operation where we're just shuffling pointers down the slice after the insertion point. But I've added a new commit that adds a `linksDirty` state to help defer the costs here. Note that it's a bit subtle because we can't call it dirty if it comes deserialized unsorted, we want to capture that state. So we have to make a decision when to sort - so `linksDirty` is done whenever you mutate the Links list, then we sort whenever we perform an operation that would need them to be in a sorted state (if they should be sorted!). But there's also operations where we have to re-sort regardless of `linksDirty`, like cloning the node, or re-serializing... Which brings up some additional questions that I've been meaning to add to the thread. Will comment separately on these.

**Comment by rvagg** [comment]
There's some additional options I wanted to flag here if anyone feels strongly enough to engage on them. IMO this currently is the right approach, but it's not the only approach. Consider the main case we're dealing with here - of a deserialized block that came with Links not properly sorted as per spec:

1. When we re-serialize the badly serialized node using this code (and all other dag-pb codecs in ipfs and ipld orgs, past and present) we end up with a properly sorted Links list, and therefore entirely new bytes (and new CID). There are some cases where this can happen in the normal flow of operating with a `ProtoNode`, one of them was demonstrated in Boost: https://github.com/filecoin-project/boost/pull/675 (although this really is a bug, I'll comment inline here about that). We have the option of preserving the badly serialized form and saying - if you want this re-serialized then you'll get back what came in. I don't think this is a good option, our codecs should behave, even with bad, but acceptable, data.
2. There is a subtle case where you're operating with a node in memory that you've built, but not serialized. If you use that node as part of a traversal and it has a form in-memory that it won't have after doing a round-trip through a codec then you might get unexpected results. (One example of this is demonstrated in https://github.com/ipld/go-ipld-prime/commit/47336e238ff0f40a419e2825f3baf45d89a77225). What we do here is ensure that the in-memory form is always consistent with what will be serialized, _unless_ it comes deserialized from a badly serialized form and hasn't been mutated. But we have the option of leaving it in the state it was constructed until we serialize it—so, e.g. if you added links in a random order to a `ProtoNode` you're building, then use it as part of a traversal (somehow) then you'll traverse it in the order you added them. We currently do this in go-ipld-prime, which we kind of have to because until a Node is serialized, it's not tied to a particular codec, all of which have different rules for sorting things (Eric often makes the argument that codecs shouldn't sort, that should be a higher-level concern, but that's another topic). I think, the appropriate choice here given the history and usage of go-merkledag is to just keep it consistent when you're building a `ProtoNode` so there are no surprises. It's always going to be dag-pb, there's no ambiguity except in the badly serialized form case.

Basically, sort your Links properly to avoid all of this.

File: coding.go [hunk 4]
```diff
@@ -132,7 +154,6 @@ func (n *ProtoNode) GetPBNode() *pb.PBNode {
 // EncodeProtobuf returns the encoded raw data version of a Node instance.
 // It may use a cached encoded version, unless the force flag is given.
 func (n *ProtoNode) EncodeProtobuf(force bool) ([]byte, error) {
-	sort.Stable(LinkSlice(n.links)) // keep links sorted
```
Comment (rvagg) [inline coding.go:135]: best to highlight the biggest culprit of the problems we're having right now - re-sorting links regardless of whether we have a cached encoded form means that just calling `RawData()` will re-sort links regardless of whether we have the original raw data or not

**Comment by rvagg** [comment]
https://github.com/ipfs/kubo/pull/9202 👍 a `dag export` sharness with a DAG made of unsorted links blocks might be a good idea at some point, but since that entire path should be using the go-ipld-prime stack I don't think it's going to add much immediate value to this PR .. especially since we're trying to migrate away from this. If it works as is now, with all tests and sharness passing then this should be pretty good.

File: node.go [hunk 5]
```diff
@@ -358,14 +403,26 @@ func (n *ProtoNode) Multihash() mh.Multihash {
...
-	return n.links
+	if n.linksDirty {
+		// there was a mutation involving links, make sure we sort
+		sort.Stable(LinkSlice(n.links))
+		n.linksDirty = false
+		n.encoded = nil
+	}
+	links := make([]*format.Link, len(n.links))
+	copy(links, n.links)
+	return links
```
Comment (Jorropo) [inline node.go:416]:
```suggestion
	return append([]*format.link(nil), n.links...)
```
This is the canonical way to make a slice copy because `make` will set the capacity to the len precisely, while `append` while extend the capacity to the underlying allocation size.
See `slices.Clone` https://pkg.go.dev/golang.org/x/exp/slices#Clone
```go
func Clone[S ~[]E, E any](s S) S {
	// Preserve nil in case it matters.
	if s == nil {
		return nil
	}
	return append(S([]E{}), s...)
}
```
You can also bump this package to `go1.18` and import `golang.org/x/exp/slices` if you want.

File: coding.go [hunk 6]
```diff
@@ -123,6 +140,11 @@ func (n *ProtoNode) GetPBNode() *pb.PBNode {
 		}
 	}
 
+	// Ensure links are sorted prior to encode, regardless of `linksDirty`. They
+	// may not have come sorted if we deserialized a badly encoded form that
+	// didn't have links already sorted.
+	sort.Stable(pbLinkSlice(pbn.Links))
```
Comment (Jorropo) [inline coding.go:146]: Wouldn't it be simpler if `Unmarshal` set the dirty bit so we would only resort the first time after unmarshal not always ?

Maybe we can't because unmarshaling is handled by other ipld code.
It's probably fine to do this go1.19's PDQSort will do a fast `O(N)` linear sweep if the data is already sorted.

File: node.go [hunk 7]
```diff
@@ -216,6 +241,10 @@ func (n *ProtoNode) Copy() format.Node {
 	if len(n.links) > 0 {
 		nnode.links = make([]*format.Link, len(n.links))
 		copy(nnode.links, n.links)
```
Comment (Jorropo) [inline node.go:243]: Same `append` note (I know you didn't modified that).
Imo just call `slices.Clone` we can now move to go1.19.

**Review by Jorropo — CHANGES_REQUESTED** [review]
I think it's good, TL;DR can you please use `slices.Clone` for slice copy ?

File: node.go [hunk 8]
```diff
@@ -358,14 +403,26 @@ func (n *ProtoNode) Multihash() mh.Multihash {
...
+	return links
 }
 
-// SetLinks replaces the node links with the given ones.
+// SetLinks replaces the node links with a copy of the provided links. Sorting
+// will be applied to the list.
 func (n *ProtoNode) SetLinks(links []*format.Link) {
-	n.links = links
+	n.links = make([]*format.Link, len(links))
+	copy(n.links, links)
```
Comment (Jorropo) [inline node.go:423]: Same copy here.

File: node.go (hunk 5 above)
Comment (rvagg) [reply node.go]: 👍 I didn't know this, but also I don't think I understand the nuance here. Is the problem that by setting the length I'm also setting the capacity, so I'm imposing a risk of a full reallocation on the user if they want to modify it themselves, whereas using `append` like this will use a default block allocation for the slice so I'm giving breathing room for modification?

File: coding.go (hunk 6 above)
Comment (rvagg) [reply coding.go:142]: Yeah, this is tricky. The main point of this PR is to allow stable inspection of the deserialized form. If we set `linksDirty` on an unmarshal, then we're going to trigger a re-sort on a number of inspection operations, losing access to the underlying order of the links as they appear in the bytes.

But on top of that, we have to decide what to do on some non-passive inspection operations, including this one (`GetPBNode()`) and `Copy()`. I've decided that in that case, we should present a _properly_ sorted form, making it not quite a 1:1 "copy" of the data, but a "copy into a correct form".

Thankfully, there are increasingly limited uses of either `GetPBNode()` or `Copy()`, and even these entire call paths, due to replacement by go-ipld-prime and go-codec-dagpb.

But, there's one path where the nuance gets a bit tricker: `marshalImmutable()` gets called by both `Marshal()` and `EncodeProtobuf()`, for the former it just wants a `[]byte` form of the Node so it just wants the links sorted. That gets done on its way out by go-codec-dagpb regardless of what it's given, so that's fine. But for `EncodeProtobuf()` it's also caching a copy of the go-ipld-prime `Node` form that's built by `marshalImmutable()`, and then that can be used to pass this `ProtoNode` off as if it's a go-ipld-prime `Node`, when `Links` field in whatever order it was given.

I've revisited my original code for that last case and have removed the sorting from  `marshalImmutable()` and put it only on `EncodeProtobuf()` only if `linksDirty` is set. That way we're fixing sorting on mutation, but capturing the Link order as they are in the bytes even for the `Node` form of this `ProtoNode`.

File: node.go (hunk 5 above)
Comment (Jorropo) [reply node.go]: Yes exactly.

The go allocator often sligthly round up, with make this extra space do nothing.
While with append it add to the capacity of the slice and can sometime save an allocation if this slice is appended to.

**Review by Jorropo — APPROVED** [review]
LGTM
